pkg/apply/applier.go

```diff
@@ -142,18 +154,130 @@ func (a *Applier) newStatusPoller() (poller.Poller, error) {
 	return polling.NewStatusPoller(c, mapper), nil
 }
 
+// previousInventories returns the infos for all the inventory objects
+// in the cluster.
 func (a *Applier) previousInventories(label, namespace string) ([]*resource.Info, error) {
 	return a.PruneOptions.RetrievePreviousInventoryObjects(label, namespace)
 }
 
+// infoHelperFactory returns a new instance of the InfoHelper.
+func (a *Applier) infoHelperFactory() info.InfoHelper {
+	return info.NewInfoHelper(a.factory, a.ApplyOptions.Namespace)
+}
+
+// readObjects reads the manifests from disk with the local flag set on
+// the Builder. This means that the resulting Info resources does not have
+// the mapping and client properties set. We are doing it this way because
+// it allows us to read in both CRDs and CRs in the same set without
+// errors.
+func (a *Applier) readObjects() ([]*resource.Info, error) {
+	r := a.factory.NewBuilder().
+		Local().
+		Unstructured().
+		Schema(a.ApplyOptions.Validator).
+		ContinueOnError().
+		NamespaceParam(a.ApplyOptions.Namespace).DefaultNamespace().
+		FilenameParam(a.ApplyOptions.EnforceNamespace, &a.ApplyOptions.DeleteOptions.FilenameOptions).
+		LabelSelectorParam(a.ApplyOptions.Selector).
+		Flatten().
+		Do()
+	if err := r.Err(); err != nil {
+		return nil, err
+	}
+
+	var infos []*resource.Info
+	err := r.Visit(func(info *resource.Info, err error) error {
+		infos = append(infos, info)
+		return nil
+	})
+	return infos, err
+}
+
+// setNamespaces verifies that
+func (a *Applier) setNamespaces(infos []*resource.Info, defaultNamespace string) error {
+	var crdInfos []*resource.Info
+
+	// find any crds in the set of resources.
+	for _, inf := range infos {
+		if solver.IsCRD(inf) {
+			crdInfos = append(crdInfos, inf)
+		}
+	}
+
+	mapper, err := a.factory.ToRESTMapper()
+	if err != nil {
+		return err
+	}
+	for _, inf := range infos {
+		accessor, _ := meta.Accessor(inf.Object)
+		// if the resource already has the namespace set, we don't
+		// need to do anything
+		if accessor.GetNamespace() != "" {
+			continue
+		}
+
+		gk := inf.Object.GetObjectKind().GroupVersionKind().GroupKind()
+		mapping, err := mapper.RESTMapping(gk)
+
+		if err != nil && !meta.IsNoMatchError(err) {
+			return err
+		}
+
+		if err == nil {
+			// If we find a mapping for the resource type in the RESTMapper,
+			// we just use it.
+			if mapping.Scope == meta.RESTScopeNamespace {
+				inf.Namespace = defaultNamespace
```

Comment:
Why are we setting to "defaultNamespace" when mapping is found? Where is "defaultNamespace" set?

Reply:
Mapping found means that we can tell whether the resource is namespace-scoped or cluster-scoped. And if it is namespace-scoped, we need to set the namespace.